content/ADR-204-comms-architecture.md

```diff
@@ -0,0 +1,620 @@
+|-------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------|
+| 0 - (PARCEL_BITS - 1)                                                                                                   | **Parcel Index**: Linearized parcel index                                                                                         |
+| PARCEL_BITS - (PARCEL_BITS + XZ_BITS - 1)                                                                               | **X Relative Position**:<br/> Uses Float Quantization:<br/>`minValue = 0`<br/>`maxValue = PARCEL_SIZE`<br/>`sizeInBits = XZ_BITS` |
+| (PARCEL_BITS + XZ_BITS) - (PARCEL_BITS + 2 * XZ_BITS - 1)                                                               | **Z Relative Position**:<br/> Uses Float Quantization:<br/>`minValue = 0`<br/>`maxValue = PARCEL_SIZE`<br/>`sizeInBits = XZ_BITS` |
+| (PARCEL_BITS + 2 * XZ_BITS) - (PARCEL_BITS + 2 * XZ_BITS + Y_BITS - 1)                                                  | **Y Absolute Position**:<br/> Uses Float Quantization:<br/>`minValue = 0`<br/>`maxValue = Y_MAX`<br/>`sizeInBits = Y_BITS`        |
+| (PARCEL_BITS + 2 * XZ_BITS + Y_BITS) - (PARCEL_BITS + 2 * XZ_BITS + Y_BITS + VELOCITY_BITS - 1)                         | **X Velocity**:<br/> Uses Velocity Compression:<br/>`minValue = 0`<br/>`maxValue = MAX_SPEED`<br/>`sizeInBits = VELOCITY_BITS`    |
+| (PARCEL_BITS + 2 * XZ_BITS + Y_BITS + VELOCITY_BITS) - (PARCEL_BITS + 2 * XZ_BITS + Y_BITS + 2 * VELOCITY_BITS - 1)     | **Y Velocity**:<br/> Uses Velocity Compression:<br/>`minValue = 0`<br/>`maxValue = MAX_SPEED`<br/>`sizeInBits = VELOCITY_BITS`    |
+| (PARCEL_BITS + 2 * XZ_BITS + Y_BITS + 2 * VELOCITY_BITS) - (PARCEL_BITS + 2 * XZ_BITS + Y_BITS + 3 * VELOCITY_BITS - 1) | **Z Velocity**:<br/> Uses Velocity Compression:<br/>`minValue = 0`<br/>`maxValue = MAX_SPEED`<br/>`sizeInBits = VELOCITY_BITS`    |
+
+**Parcel Linearization**
+
+```csharp
+        public static class GenesisCityData
+        {
+            public static readonly Vector2Int MIN_PARCEL = -150 * Vector2Int.one;
+            public static readonly Vector2Int MAX_PARCEL = new (163, 158);
+        }
+        
+        public int MinX => GenesisCityData.MIN_PARCEL.x - terrainData.borderPadding;
+        public int MinY => GenesisCityData.MIN_PARCEL.y - terrainData.borderPadding;
+        public int MaxX => GenesisCityData.MAX_PARCEL.x + terrainData.borderPadding;
+        public int MaxY => GenesisCityData.MAX_PARCEL.y + terrainData.borderPadding;
+        
+        public int Encode(Vector2Int parcel) =>
+            parcel.x - MinX + ((parcel.y - MinY) * width);
+
+        public Vector2Int Decode(int index) =>
+            new ((index % width) + MinX, (index / width) + MinY);
+```
+
+`terrainData.borderPadding = 2` in the new client.
```

Comment:
Anyway, we should talk about another way to set up the relative parcel you are. Setting 17 bits to this index is a hard-limit that we already are close with Genesis City.

Reply:
Let's try to avoid referencing `new client` and make it more generic

Reply:
> If we want to ensure that that different clients can properly decode the message we need to have this included in ADR.
I agree that all the information should be here in the ADR. But we can approach an identical implementation even if we remove the genesis city shape&size from here.

Setting different origins for player to refer the relative position is a good approach. Then, we can take advantage of that for sending less data. But I think it's a bad idea to put Genesis City shape and size here to index parcels. I also think 17bits is a hard-limit and really close to what we currently have for Genesis-City.

I can add the idea of using another message to send the references/settings instead of hard-coding it.

Reply:
@leanmendoza for me the class serves as a reference to GenesisCity data and is not a canonical model for any metaverse. I believe it functions well as an implementation and can be adapted for other use cases. How would you suggest documenting this as part of the current implementation?

Reply:
@pentreathm 
What the The data transport relies on this index, it's not only a reference. The first 0-16 bits are the index based on this class, if you call Decode(0) => (-152, -152)

The indexes parameters could easily be a different message for setting up the comms, read my other comment https://github.com/decentraland/adr/pull/272#discussion_r1865581408 the idea of MovementHeader (sent only once per player maybe)

Reply:
```
            public static readonly Vector2Int MIN_PARCEL = -150 * Vector2Int.one;
            public static readonly Vector2Int MAX_PARCEL = new (163, 158);
```

It should not be hardcoded, we should get that data from one somewhere.
It can be in the `/about` or a MovementHeader like @leanmendoza is proposing.

The Protocol should be agnostic to Genesis City.
Genesis City is a protocol implementation, not the protocol.

Reply:
I see, we could expose it in the /about under `comms` if needed. @mikhail-dcl what do you think about these options?

Reply:
We have that data available in the /about endpoint right now.

Example: https://realm-provider-ea.decentraland.org/main/about

We can process the sizes in the `configurations.map.sizes` and get the minimum bottom, maximum top, minimum left, and maximum right values. Then, we can use these values.

```
...
  "configurations": {
    "map": {
      ...
      "sizes": [
        { "left": -150, "top": 150, "right": 150, "bottom": -150 },
        { "left": 62, "top": 158, "right": 162, "bottom": 151 },
        { "left": 151, "top": 150, "right": 163, "bottom": 59 }
      ],
      ...
    }
  },
```
If you query Genesis Plaza, you will get:
```
Maximum right: 163
Minimum left: -150
Maximum top: 158
Minimum bottom: -150
```

which is the same data used in your definition:
```
            public static readonly Vector2Int MIN_PARCEL = -150 * Vector2Int.one;
            public static readonly Vector2Int MAX_PARCEL = new (163, 158);
```

here is the ADR of that spec: https://adr.decentraland.org/adr/ADR-250

Reply:
We decided to move parcel information to `Metadata`: it will no longer be compressed and will be set only when the parcel is changed. Thus, your concern will be covered and no extra info is needed.

We need to try it out first in E@ to understand how reliable it is before moving on.

Reply:
@mikhail-dcl could you please update the ADR so that we can resolve the comments and merge it?